snow/engine/snowman/straggler_detect.go

```diff
@@ -0,0 +1,331 @@
+	ourLastAcceptedBlock := sd.lastAccepted()
+
+	netInfo, err := sd.getNetworkInfo(ourLastAcceptedBlock)
+	if err != nil {
+		return snapshot{}, false
+	}
+
+	if !sd.validateNetInfo(netInfo) {
+		return snapshot{}, false
+	}
+
+	snap := snapshot{
+		nodeWeightsToBlocks:  netInfo.nodeWeightToLastAccepted,
+		totalValidatorWeight: netInfo.totalValidatorWeight,
+	}
+
+	if sd.haveWeFailedCatchingUp(snap) {
+		return snap, true
+	}
+
+	return snapshot{}, false
+}
+
+type netInfo struct {
+	connStakePercent                      float64
+	totalPendingStake                     uint64
+	totalValidatorWeight                  uint64
+	totalWeightWeKnowItsLastAcceptedBlock uint64
+	nodeWeightToLastAccepted              nodeWeightsToBlocks
+}
+
+func (sd *stragglerDetector) getNetworkInfo(ourLastAcceptedBlock ids.ID) (netInfo, error) {
+	var res netInfo
+
+	res.connStakePercent = sd.connectedPercent()
+
+	validators := nodeWeights(sd.connectedValidators().List())
+	nodeWeightToLastAccepted := make(nodeWeightsToBlocks, len(validators))
+
+	for _, vdr := range validators {
+		lastAccepted, ok := sd.lastAcceptedByNodeID(vdr.Node)
+		if !ok {
+			continue
+		}
+		nodeWeightToLastAccepted[vdr] = lastAccepted
+	}
+
+	totalValidatorWeight, err := validators.totalWeight()
+	if err != nil {
+		sd.log.Error("Failed computing total weight", zap.Error(err))
+		return netInfo{}, err
+	}
+	res.totalValidatorWeight = totalValidatorWeight
+
+	totalWeightWeKnowItsLastAcceptedBlock, err := nodeWeightToLastAccepted.totalWeight()
+	if err != nil {
+		sd.log.Error("Failed computing total weight", zap.Error(err))
+		return netInfo{}, err
+	}
+	res.totalWeightWeKnowItsLastAcceptedBlock = totalWeightWeKnowItsLastAcceptedBlock
+
+	prevLastAcceptedCount := len(nodeWeightToLastAccepted)
+
+	// Ensure we have collected last accepted blocks that are not our own last accepted block.
+	for nodeWeight, lastAccepted := range nodeWeightToLastAccepted {
+		if ourLastAcceptedBlock.Compare(lastAccepted) == 0 {
+			delete(nodeWeightToLastAccepted, nodeWeight)
+		}
+	}
+
+	res.nodeWeightToLastAccepted = nodeWeightToLastAccepted
+
+	totalPendingStake, err := nodeWeightToLastAccepted.totalWeight()
+	if err != nil {
+		sd.log.Error("Failed computing total weight", zap.Error(err))
+		return netInfo{}, err
+	}
+
+	res.totalPendingStake = totalPendingStake
+
+	sd.log.Trace("Excluding nodes with our own height", zap.Int("prev", prevLastAcceptedCount), zap.Uint64("new", totalPendingStake))
+
+	return res, nil
+}
+
+type snapshot struct {
+	totalValidatorWeight uint64
+	nodeWeightsToBlocks  nodeWeightsToBlocks
+}
+
+func (s snapshot) isEmpty() bool {
+	return s.totalValidatorWeight == 0 || len(s.nodeWeightsToBlocks) == 0
+}
+
+type nodeWeightsToBlocks map[ids.NodeWeight]ids.ID
+
+func (nwb nodeWeightsToBlocks) totalWeight() (uint64, error) {
+	return nodeWeights(maps.Keys(nwb)).totalWeight()
+}
+
+func dropHeight(f func() (ids.ID, uint64)) func() ids.ID {
```

Comment:
this function is a bit difficult to follow, and not sure if we %100 need it.

Reply:
can we add comment or prefably drop the function? IMHO we should be using tracker.Accepted interface anyway

Reply:
I added a comment.

Reply:
Could we remove this function? Seems like overkill to use in a single place